homeassistant/components/automation/time.py

```diff
@@ -1,44 +1,118 @@
 
-    removes = [
-        async_track_time_change(
-            hass,
-            time_automation_listener,
-            hour=at_time.hour,
-            minute=at_time.minute,
-            second=at_time.second,
-        )
-        for at_time in at_times
-    ]
+    @callback
+    def update_entity_trigger(entity_id, old_state=None, new_state=None):
+        # If a listener was already set up for entity, remove it.
+        remove = entities.get(entity_id)
+        if remove:
+            remove()
+            removes.remove(remove)
+            remove = None
+
+        # Check state of entity. If valid, set up a listener.
+        if new_state:
+            has_date = new_state.attributes["has_date"]
+            if has_date:
+                year = new_state.attributes["year"]
+                month = new_state.attributes["month"]
+                day = new_state.attributes["day"]
+            has_time = new_state.attributes["has_time"]
+            if has_time:
+                hour = new_state.attributes["hour"]
+                minute = new_state.attributes["minute"]
+                second = new_state.attributes["second"]
+            else:
+                # If no time then use midnight.
+                hour = minute = second = 0
+
+            if has_date:
+                # If input_datetime has date, then track point in time.
+                trigger_dt = dt_util.DEFAULT_TIME_ZONE.localize(
+                    datetime(year, month, day, hour, minute, second)
+                )
+                # Only set up listener if time is now or in the future.
+                if trigger_dt >= dt_util.now():
+                    remove = async_track_point_in_time(
+                        hass, time_automation_listener, trigger_dt
+                    )
+            elif has_time:
+                # Else if it has time, then track time change.
+                remove = async_track_time_change(
+                    hass,
+                    time_automation_listener,
+                    hour=hour,
+                    minute=minute,
+                    second=second,
+                )
+
+        # Was a listener set up?
+        if remove:
+            removes.append(remove)
+
+        entities[entity_id] = remove
+
+    for at_time in config[CONF_AT]:
+        if isinstance(at_time, str):
+            # input_datetime entity
+            update_entity_trigger(at_time, new_state=hass.states.get(at_time))
+        else:
+            # datetime.time
+            removes.append(
+                async_track_time_change(
+                    hass,
+                    time_automation_listener,
+                    hour=at_time.hour,
+                    minute=at_time.minute,
+                    second=at_time.second,
+                )
+            )
+
+    # Track state changes of any entities.
+    removes.append(
+        async_track_state_change(hass, list(entities), update_entity_trigger)
```

Comment:
Shouldn't this be using the faster `async_track_state_change_event` instead?

Reply:
`async_track_state_change` will use `async_track_state_change_event`, but will also convert the event into new & old states, which `update_entity_trigger` needs because it is also called in another place where an event object isn't available. I think it's fine the way it is.